File: cli/internal/history.go
package internal

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

const historyVersion = 1

type TaskSnapshot struct {
	ID         int    `json:"id"`
	Title      string `json:"title"`
	Duration   int    `json:"duration"`
	Status     string `json:"status"`
	Deadline   string `json:"deadline,omitempty"`
	Visibility []int  `json:"visibility,omitempty"`
}

type HistoryDay struct {
	Date      string         `json:"date"`
	UpdatedAt int64          `json:"updated_at,omitempty"`
	Tasks     []TaskSnapshot `json:"tasks"`
}

type HistoryEvent struct {
	Timestamp  int64  `json:"timestamp"`
	Date       string `json:"date"`
	Type       string `json:"type"`
	TaskID     int    `json:"task_id"`
	Title      string `json:"title"`
	FromStatus string `json:"from_status,omitempty"`
	ToStatus   string `json:"to_status,omitempty"`
	Duration   int    `json:"duration,omitempty"`
	Deadline   string `json:"deadline,omitempty"`
}

type History struct {
	Version   int            `json:"version"`
	UpdatedAt int64          `json:"updated_at,omitempty"`
	Days      []HistoryDay   `json:"days,omitempty"`
	Events    []HistoryEvent `json:"events,omitempty"`
}

type DailyStats struct {
	Date            string  `json:"date"`
	TaskCount       int     `json:"task_count"`
	TodoCount       int     `json:"todo_count"`
	DoneCount       int     `json:"done_count"`
	SkippedCount    int     `json:"skipped_count"`
	TodoDuration    int     `json:"todo_duration"`
	DoneDuration    int     `json:"done_duration"`
	SkippedDuration int     `json:"skipped_duration"`
	CompletionRate  float64 `json:"completion_rate"`
}

type TaskFrequencyStats struct {
	TaskID          int     `json:"task_id"`
	Title           string  `json:"title"`
	RecordedDays    int     `json:"recorded_days"`
	TodoDays        int     `json:"todo_days"`
	DoneDays        int     `json:"done_days"`
	SkippedDays     int     `json:"skipped_days"`
	CompletionRate  float64 `json:"completion_rate"`
	TotalDuration   int     `json:"total_duration"`
	DoneDuration    int     `json:"done_duration"`
	SkippedDuration int     `json:"skipped_duration"`
}

type StatsSummary struct {
	From            string               `json:"from"`
	To              string               `json:"to"`
	RecordedDays    int                  `json:"recorded_days"`
	TaskCount       int                  `json:"task_count"`
	TodoCount       int                  `json:"todo_count"`
	DoneCount       int                  `json:"done_count"`
	SkippedCount    int                  `json:"skipped_count"`
	TodoDuration    int                  `json:"todo_duration"`
	DoneDuration    int                  `json:"done_duration"`
	SkippedDuration int                  `json:"skipped_duration"`
	CompletionRate  float64              `json:"completion_rate"`
	Daily           []DailyStats         `json:"daily"`
	Tasks           []TaskFrequencyStats `json:"tasks"`
}

func HistoryPath(dataPath string) string {
	ext := filepath.Ext(dataPath)
	if ext == "" {
		return dataPath + ".history.json"
	}
	base := strings.TrimSuffix(dataPath, ext)
	return base + ".history" + ext
}

func LoadHistory(dataPath string) (History, error) {
	path := HistoryPath(dataPath)
	b, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return History{Version: historyVersion}, nil
		}
		return History{}, err
	}

	var history History
	if err := json.Unmarshal(b, &history); err != nil {
		return History{}, err
	}
	if history.Version == 0 {
		history.Version = historyVersion
	}
	sortHistory(&history)
	return history, nil
}

func SaveHistory(dataPath string, history History) error {
	history.Version = historyVersion
	history.UpdatedAt = time.Now().UnixMilli()
	sortHistory(&history)

	b, err := json.MarshalIndent(history, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(HistoryPath(dataPath), b, 0o600)
}

func HistoryWithCurrentSnapshot(history History, data Data, now int64) History {
	normalized := NormalizeData(data)
	if now == 0 {
		now = time.Now().UnixMilli()
	}
	if history.Version == 0 {
		history.Version = historyVersion
	}
	if upsertHistoryDay(&history, normalized, now) {
		history.UpdatedAt = now
	}
	sortHistory(&history)
	return history
}

func HistoryContentEqual(a, b History) bool {
	if a.Version == 0 {
		a.Version = historyVersion
	}
	if b.Version == 0 {
		b.Version = historyVersion
	}
	sortHistory(&a)
	sortHistory(&b)
	if a.Version != b.Version || len(a.Days) != len(b.Days) || len(a.Events) != len(b.Events) {
		return false
	}
	for i := range a.Days {
		if !historyDayEqual(a.Days[i], b.Days[i]) {
			return false
		}
	}
	for i := range a.Events {
		if a.Events[i] != b.Events[i] {
			return false
		}
	}
	return true
}

func MergeHistories(local, remote History) History {
	if local.Version == 0 {
		local.Version = historyVersion
	}
	if remote.Version == 0 {
		remote.Version = historyVersion
	}
	merged := History{
		Version: max(local.Version, remote.Version),
	}

	dayMap := map[string]HistoryDay{}
	for _, day := range append(local.Days, remote.Days...) {
		existing, ok := dayMap[day.Date]
		if !ok || day.UpdatedAt > existing.UpdatedAt || (day.UpdatedAt == existing.UpdatedAt && len(day.Tasks) >= len(existing.Tasks)) {
			copyDay := HistoryDay{
				Date:      day.Date,
				UpdatedAt: day.UpdatedAt,
				Tasks:     append([]TaskSnapshot(nil), day.Tasks...),
			}
			dayMap[day.Date] = copyDay
		}
	}
	for _, day := range dayMap {
		merged.Days = append(merged.Days, day)
	}

	eventMap := map[string]HistoryEvent{}
	for _, event := range append(local.Events, remote.Events...) {
		key := strings.Join([]string{
			strconv.FormatInt(event.Timestamp, 10),
			event.Date,
			event.Type,
			strconv.Itoa(event.TaskID),
			event.Title,
			event.FromStatus,
			event.ToStatus,
			strconv.Itoa(event.Duration),
			event.Deadline,
		}, "|")
		if _, ok := eventMap[key]; !ok {
			eventMap[key] = event
		}
	}
	for _, event := range eventMap {
		merged.Events = append(merged.Events, event)
	}

	if local.UpdatedAt > merged.UpdatedAt {
		merged.UpdatedAt = local.UpdatedAt
	}
	if remote.UpdatedAt > merged.UpdatedAt {
		merged.UpdatedAt = remote.UpdatedAt
	}
	sortHistory(&merged)
	return merged
}

func EnsureHistorySnapshot(dataPath string, data Data) error {
	history, err := LoadHistory(dataPath)
	if err != nil {
		return err
	}
	if upsertHistoryDay(&history, NormalizeData(data), time.Now().UnixMilli()) {
		return SaveHistory(dataPath, history)
	}
	return nil
}

func SaveDataWithHistory(dataPath string, before, after Data) error {
	normalizedBefore := NormalizeData(before)
	normalizedAfter := NormalizeData(after)

	if err := SaveData(dataPath, normalizedAfter); err != nil {
		return err
	}

	history, err := LoadHistory(dataPath)
	if err != nil {
		return err
	}

	now := time.Now().UnixMilli()
	initialEventCount := len(history.Events)
	changed := false
	if normalizedBefore.LastReset != "" {
		changed = upsertHistoryDay(&history, normalizedBefore, now) || changed
	}
	appendHistoryEvents(&history, normalizedBefore, normalizedAfter, now)
	changed = upsertHistoryDay(&history, normalizedAfter, now) || changed

	if !changed && len(history.Events) == initialEventCount {
		return nil
	}
	return SaveHistory(dataPath, history)
}

func BuildStats(dataPath string, current Data, from, to string) (StatsSummary, error) {
	history, err := LoadHistory(dataPath)
	if err != nil {
		return StatsSummary{}, err
	}

	withSnapshot := HistoryWithCurrentSnapshot(history, current, time.Now().UnixMilli())
	if !HistoryContentEqual(withSnapshot, history) {
		if err := SaveHistory(dataPath, withSnapshot); err != nil {
			return StatsSummary{}, err
		}
	}

	return AggregateStats(withSnapshot, from, to), nil
}

func AggregateStats(history History, from, to string) StatsSummary {
	sortHistory(&history)

	filtered := make([]HistoryDay, 0, len(history.Days))
	for _, day := range history.Days {
		if from != "" && day.Date < from {
			continue
		}
		if to != "" && day.Date > to {
			continue
		}
		filtered = append(filtered, day)
	}

	summary := StatsSummary{
		From:         from,
		To:           to,
		RecordedDays: len(filtered),
		Daily:        make([]DailyStats, 0, len(filtered)),
	}

	type taskAccumulator struct {
		TaskFrequencyStats
	}

	taskMap := map[int]*taskAccumulator{}
	for _, day := range filtered {
		daily := DailyStats{Date: day.Date}
		for _, task := range day.Tasks {
			daily.TaskCount++
			summary.TaskCount++

			acc, ok := taskMap[task.ID]
			if !ok {
				acc = &taskAccumulator{
					TaskFrequencyStats: TaskFrequencyStats{
						TaskID: task.ID,
						Title:  task.Title,
					},
				}
				taskMap[task.ID] = acc
			}
			acc.Title = task.Title
			acc.RecordedDays++
			acc.TotalDuration += task.Duration

			switch task.Status {
			case "done":
				daily.DoneCount++
				daily.DoneDuration += task.Duration
				summary.DoneCount++
				summary.DoneDuration += task.Duration
				acc.DoneDays++
				acc.DoneDuration += task.Duration
			case "skipped":
				daily.SkippedCount++
				daily.SkippedDuration += task.Duration
				summary.SkippedCount++
				summary.SkippedDuration += task.Duration
				acc.SkippedDays++
				acc.SkippedDuration += task.Duration
			default:
				daily.TodoCount++
				daily.TodoDuration += task.Duration
				summary.TodoCount++
				summary.TodoDuration += task.Duration
				acc.TodoDays++
			}
		}
		if daily.TaskCount > 0 {
			daily.CompletionRate = float64(daily.DoneCount) / float64(daily.TaskCount)
		}
		summary.Daily = append(summary.Daily, daily)
	}

	if summary.TaskCount > 0 {
		summary.CompletionRate = float64(summary.DoneCount) / float64(summary.TaskCount)
	}

	for _, acc := range taskMap {
		if acc.RecordedDays > 0 {
			acc.CompletionRate = float64(acc.DoneDays) / float64(acc.RecordedDays)
		}
		summary.Tasks = append(summary.Tasks, acc.TaskFrequencyStats)
	}

	sort.Slice(summary.Tasks, func(i, j int) bool {
		if summary.Tasks[i].DoneDays == summary.Tasks[j].DoneDays {
			if summary.Tasks[i].RecordedDays == summary.Tasks[j].RecordedDays {
				return summary.Tasks[i].TaskID < summary.Tasks[j].TaskID
			}
			return summary.Tasks[i].RecordedDays > summary.Tasks[j].RecordedDays
		}
		return summary.Tasks[i].DoneDays > summary.Tasks[j].DoneDays
	})

	return summary
}

func sortHistory(history *History) {
	sort.Slice(history.Days, func(i, j int) bool {
		return history.Days[i].Date < history.Days[j].Date
	})
	sort.Slice(history.Events, func(i, j int) bool {
		if history.Events[i].Timestamp == history.Events[j].Timestamp {
			if history.Events[i].TaskID == history.Events[j].TaskID {
				return history.Events[i].Type < history.Events[j].Type
			}
			return history.Events[i].TaskID < history.Events[j].TaskID
		}
		return history.Events[i].Timestamp < history.Events[j].Timestamp
	})
}

func upsertHistoryDay(history *History, data Data, now int64) bool {
	day := HistoryDay{
		Date:      data.LastReset,
		UpdatedAt: now,
		Tasks:     snapshotsForVisibleTasks(data.Tasks, data.LastReset),
	}

	for i := range history.Days {
		if history.Days[i].Date == day.Date {
			if historyDayEqual(history.Days[i], day) {
				return false
			}
			history.Days[i] = day
			return true
		}
	}

	history.Days = append(history.Days, day)
	return true
}

func historyDayEqual(a, b HistoryDay) bool {
	if a.Date != b.Date || len(a.Tasks) != len(b.Tasks) {
		return false
	}
	for i := range a.Tasks {
		ta, tb := a.Tasks[i], b.Tasks[i]
		if ta.ID != tb.ID || ta.Title != tb.Title || ta.Duration != tb.Duration ||
			ta.Status != tb.Status || ta.Deadline != tb.Deadline {
			return false
		}
		if len(ta.Visibility) != len(tb.Visibility) {
			return false
		}
		for j := range ta.Visibility {
			if ta.Visibility[j] != tb.Visibility[j] {
				return false
			}
		}
	}
	return true
}

func snapshotsForTasks(tasks []Task) []TaskSnapshot {
	snapshots := make([]TaskSnapshot, 0, len(tasks))
	for _, task := range tasks {
		snapshots = append(snapshots, TaskSnapshot{
			ID:         task.ID,
			Title:      task.Title,
			Duration:   task.Duration,
			Status:     task.Status,
			Deadline:   task.Deadline,
			Visibility: task.Visibility,
		})
	}
	sort.Slice(snapshots, func(i, j int) bool {
		return snapshots[i].ID < snapshots[j].ID
	})
	return snapshots
}

// snapshotsForVisibleTasks returns snapshots only for tasks visible on the
// given date string (YYYY-MM-DD). Falls back to all tasks on parse error.
func snapshotsForVisibleTasks(tasks []Task, date string) []TaskSnapshot {
	weekday, err := WeekdayFromDate(date)
	if err != nil {
		return snapshotsForTasks(tasks)
	}
	visible := VisibleTasksOn(tasks, weekday)
	return snapshotsForTasks(visible)
}

func appendHistoryEvents(history *History, before, after Data, now int64) {
	beforeMap := map[int]Task{}
	afterMap := map[int]Task{}

	for _, task := range before.Tasks {
		beforeMap[task.ID] = task
	}
	for _, task := range after.Tasks {
		afterMap[task.ID] = task
	}

	for id, previous := range beforeMap {
		next, ok := afterMap[id]
		if !ok {
			history.Events = append(history.Events, HistoryEvent{
				Timestamp:  now,
				Date:       before.LastReset,
				Type:       "task_deleted",
				TaskID:     previous.ID,
				Title:      previous.Title,
				FromStatus: previous.Status,
				Duration:   previous.Duration,
				Deadline:   previous.Deadline,
			})
			continue
		}

		if previous.Status != next.Status {
			history.Events = append(history.Events, HistoryEvent{
				Timestamp:  now,
				Date:       after.LastReset,
				Type:       "status_changed",
				TaskID:     next.ID,
				Title:      next.Title,
				FromStatus: previous.Status,
				ToStatus:   next.Status,
				Duration:   next.Duration,
				Deadline:   next.Deadline,
			})
		}

		if previous.Title != next.Title || previous.Duration != next.Duration || previous.Deadline != next.Deadline {
			history.Events = append(history.Events, HistoryEvent{
				Timestamp: now,
				Date:      after.LastReset,
				Type:      "task_updated",
				TaskID:    next.ID,
				Title:     next.Title,
				ToStatus:  next.Status,
				Duration:  next.Duration,
				Deadline:  next.Deadline,
			})
		}
	}

	for id, task := range afterMap {
		if _, ok := beforeMap[id]; ok {
			continue
		}
		history.Events = append(history.Events, HistoryEvent{
			Timestamp: now,
			Date:      after.LastReset,
			Type:      "task_added",
			TaskID:    task.ID,
			Title:     task.Title,
			ToStatus:  task.Status,
			Duration:  task.Duration,
			Deadline:  task.Deadline,
		})
	}
}

File: mobile/src/services/history.ts
import { cloneData, normalizeData, resetIfNeeded } from './data';
import type {
  Data,
  DailyStats,
  History,
  HistoryDay,
  HistoryEvent,
  StatsSummary,
  Task,
  TaskFrequencyStats,
  TaskSnapshot,
} from '../types';

const HISTORY_VERSION = 1;

type TaskAccumulator = TaskFrequencyStats;

export function emptyHistory(): History {
  return {
    version: HISTORY_VERSION,
    days: [],
    events: [],
  };
}

export function normalizeHistory(input?: Partial<History> | null): History {
  const history: History = {
    version: input?.version || HISTORY_VERSION,
    updated_at: input?.updated_at,
    days: (input?.days ?? []).map((day) => ({
      date: day.date || '',
      updated_at: day.updated_at,
      tasks: (day.tasks ?? []).map((task) => ({
        id: task.id,
        title: task.title,
        duration: task.duration,
        status: task.status,
        deadline: task.deadline,
      })),
    })),
    events: (input?.events ?? []).map((event) => ({
      timestamp: event.timestamp,
      date: event.date || '',
      type: event.type,
      task_id: event.task_id,
      title: event.title,
      from_status: event.from_status,
      to_status: event.to_status,
      duration: event.duration,
      deadline: event.deadline,
    })),
  };

  sortHistory(history);
  return history;
}

export function ensureHistorySnapshot(historyInput: History, current: Data, now = Date.now()): History {
  const history = normalizeHistory(historyInput);
  const changed = upsertHistoryDay(history, normalizeData(cloneData(current)), now);
  if (changed) {
    history.updated_at = now;
  }
  sortHistory(history);
  return history;
}

export function historyContentEqual(aInput: History, bInput: History): boolean {
  const a = normalizeHistory(aInput);
  const b = normalizeHistory(bInput);

  if (a.version !== b.version || a.days.length !== b.days.length || a.events.length !== b.events.length) {
    return false;
  }

  for (let i = 0; i < a.days.length; i += 1) {
    if (!historyDayEqual(a.days[i], b.days[i])) {
      return false;
    }
  }

  for (let i = 0; i < a.events.length; i += 1) {
    const left = a.events[i];
    const right = b.events[i];
    if (
      left.timestamp !== right.timestamp ||
      left.date !== right.date ||
      left.type !== right.type ||
      left.task_id !== right.task_id ||
      left.title !== right.title ||
      left.from_status !== right.from_status ||
      left.to_status !== right.to_status ||
      left.duration !== right.duration ||
      left.deadline !== right.deadline
    ) {
      return false;
    }
  }

  return true;
}

export function mergeHistories(localInput: History, remoteInput: History): History {
  const local = normalizeHistory(localInput);
  const remote = normalizeHistory(remoteInput);

  const dayMap = new Map<string, HistoryDay>();
  for (const day of [...local.days, ...remote.days]) {
    const existing = dayMap.get(day.date);
    if (!existing) {
      dayMap.set(day.date, { ...day, tasks: day.tasks.map((task) => ({ ...task })) });
      continue;
    }

    const existingUpdatedAt = existing.updated_at || 0;
    const nextUpdatedAt = day.updated_at || 0;
    if (nextUpdatedAt > existingUpdatedAt || (nextUpdatedAt === existingUpdatedAt && day.tasks.length >= existing.tasks.length)) {
      dayMap.set(day.date, { ...day, tasks: day.tasks.map((task) => ({ ...task })) });
    }
  }

  const eventMap = new Map<string, HistoryEvent>();
  for (const event of [...local.events, ...remote.events]) {
    const key = [
      event.timestamp,
      event.date,
      event.type,
      event.task_id,
      event.title,
      event.from_status || '',
      event.to_status || '',
      event.duration || 0,
      event.deadline || '',
    ].join('|');
    if (!eventMap.has(key)) {
      eventMap.set(key, { ...event });
    }
  }

  const merged: History = {
    version: Math.max(local.version || HISTORY_VERSION, remote.version || HISTORY_VERSION),
    updated_at: Math.max(local.updated_at || 0, remote.updated_at || 0),
    days: Array.from(dayMap.values()),
    events: Array.from(eventMap.values()),
  };

  sortHistory(merged);
  return merged;
}

export function recordDataChange(historyInput: History, before: Data, after: Data, now = Date.now()): History {
  const history = normalizeHistory(historyInput);
  const normalizedBefore = normalizeData(cloneData(before));
  const normalizedAfter = normalizeData(cloneData(after));

  let changed = false;
  const initialEventCount = history.events.length;

  if (normalizedBefore.last_reset) {
    changed = upsertHistoryDay(history, normalizedBefore, now) || changed;
  }

  appendHistoryEvents(history, normalizedBefore, normalizedAfter, now);
  changed = upsertHistoryDay(history, normalizedAfter, now) || changed;

  if (!changed && history.events.length === initialEventCount) {
    return history;
  }

  history.updated_at = now;
  sortHistory(history);
  return history;
}

export function applyDailyResetWithHistory(dataInput: Data, historyInput: History, now = Date.now()): {
  data: Data;
  history: History;
  changed: boolean;
} {
  const before = normalizeData(cloneData(dataInput));
  const after = resetIfNeeded(cloneData(before));
  if (before.last_reset === after.last_reset) {
    return { data: after, history: normalizeHistory(historyInput), changed: false };
  }

  const history = recordDataChange(historyInput, before, after, now);
  return { data: after, history, changed: true };
}

export function buildStatsSummary(historyInput: History, current: Data, from: string, to: string): StatsSummary {
  const history = ensureHistorySnapshot(historyInput, current, Date.now());
  return aggregateStats(history, from, to);
}

export function aggregateStats(historyInput: History, from: string, to: string): StatsSummary {
  const history = normalizeHistory(historyInput);
  const filtered = history.days.filter((day) => {
    if (from && day.date < from) {
      return false;
    }
    if (to && day.date > to) {
      return false;
    }
    return true;
  });

  const summary: StatsSummary = {
    from,
    to,
    recorded_days: filtered.length,
    task_count: 0,
    todo_count: 0,
    done_count: 0,
    skipped_count: 0,
    todo_duration: 0,
    done_duration: 0,
    skipped_duration: 0,
    completion_rate: 0,
    daily: [],
    tasks: [],
  };

  const taskMap = new Map<number, TaskAccumulator>();

  for (const day of filtered) {
    const daily: DailyStats = {
      date: day.date,
      task_count: 0,
      todo_count: 0,
      done_count: 0,
      skipped_count: 0,
      todo_duration: 0,
      done_duration: 0,
      skipped_duration: 0,
      completion_rate: 0,
    };

    for (const task of day.tasks) {
      daily.task_count += 1;
      summary.task_count += 1;

      const currentTask = taskMap.get(task.id) ?? {
        task_id: task.id,
        title: task.title,
        recorded_days: 0,
        todo_days: 0,
        done_days: 0,
        skipped_days: 0,
        completion_rate: 0,
        total_duration: 0,
        done_duration: 0,
        skipped_duration: 0,
      };

      currentTask.title = task.title;
      currentTask.recorded_days += 1;
      currentTask.total_duration += task.duration;
      taskMap.set(task.id, currentTask);

      switch (task.status) {
        case 'done':
          daily.done_count += 1;
          daily.done_duration += task.duration;
          summary.done_count += 1;
          summary.done_duration += task.duration;
          currentTask.done_days += 1;
          currentTask.done_duration += task.duration;
          break;
        case 'skipped':
          daily.skipped_count += 1;
          daily.skipped_duration += task.duration;
          summary.skipped_count += 1;
          summary.skipped_duration += task.duration;
          currentTask.skipped_days += 1;
          currentTask.skipped_duration += task.duration;
          break;
        default:
          daily.todo_count += 1;
          daily.todo_duration += task.duration;
          summary.todo_count += 1;
          summary.todo_duration += task.duration;
          currentTask.todo_days += 1;
          break;
      }
    }

    if (daily.task_count > 0) {
      daily.completion_rate = daily.done_count / daily.task_count;
    }

    summary.daily.push(daily);
  }

  if (summary.task_count > 0) {
    summary.completion_rate = summary.done_count / summary.task_count;
  }

  summary.tasks = Array.from(taskMap.values())
    .map((task) => ({
      ...task,
      completion_rate: task.recorded_days > 0 ? task.done_days / task.recorded_days : 0,
    }))
    .sort((a, b) => {
      if (a.done_days === b.done_days) {
        if (a.recorded_days === b.recorded_days) {
          return a.task_id - b.task_id;
        }
        return b.recorded_days - a.recorded_days;
      }
      return b.done_days - a.done_days;
    });

  return summary;
}

function sortHistory(history: History): void {
  history.days.sort((a, b) => a.date.localeCompare(b.date));
  history.events.sort((a, b) => {
    if (a.timestamp === b.timestamp) {
      if (a.task_id === b.task_id) {
        return a.type.localeCompare(b.type);
      }
      return a.task_id - b.task_id;
    }
    return a.timestamp - b.timestamp;
  });
}

function upsertHistoryDay(history: History, data: Data, now: number): boolean {
  const day: HistoryDay = {
    date: data.last_reset,
    updated_at: now,
    tasks: snapshotsForVisibleTasks(data.tasks, data.last_reset),
  };

  const index = history.days.findIndex((entry) => entry.date === day.date);
  if (index >= 0) {
    if (historyDayEqual(history.days[index], day)) {
      return false;
    }
    history.days[index] = day;
    return true;
  }

  history.days.push(day);
  return true;
}

function historyDayEqual(a: HistoryDay, b: HistoryDay): boolean {
  if (a.date !== b.date || a.tasks.length !== b.tasks.length) {
    return false;
  }

  return a.tasks.every((task, index) => {
    const other = b.tasks[index];
    return (
      task.id === other.id &&
      task.title === other.title &&
      task.duration === other.duration &&
      task.status === other.status &&
      task.deadline === other.deadline
    );
  });
}

// Returns snapshots only for tasks visible on the given date (YYYY-MM-DD).
// Falls back to all tasks if the date can't be parsed.
function snapshotsForVisibleTasks(tasks: Task[], date: string): TaskSnapshot[] {
  const parsed = new Date(`${date}T00:00:00`);
  if (isNaN(parsed.getTime())) {
    return snapshotsForTasks(tasks);
  }
  const weekday = parsed.getDay(); // 0=Sun … 6=Sat
  const visible = tasks.filter(
    (t) => !t.visibility || t.visibility.length === 0 || t.visibility.includes(weekday),
  );
  return snapshotsForTasks(visible);
}

function snapshotsForTasks(tasks: Task[]): TaskSnapshot[] {
  return tasks
    .map((task) => ({
      id: task.id,
      title: task.title,
      duration: task.duration,
      status: task.status,
      deadline: task.deadline,
    }))
    .sort((a, b) => a.id - b.id);
}

function appendHistoryEvents(history: History, before: Data, after: Data, now: number): void {
  const beforeMap = new Map(before.tasks.map((task) => [task.id, task]));
  const afterMap = new Map(after.tasks.map((task) => [task.id, task]));

  for (const [id, previous] of beforeMap) {
    const next = afterMap.get(id);
    if (!next) {
      history.events.push({
        timestamp: now,
        date: before.last_reset,
        type: 'task_deleted',
        task_id: previous.id,
        title: previous.title,
        from_status: previous.status,
        duration: previous.duration,
        deadline: previous.deadline,
      });
      continue;
    }

    if (previous.status !== next.status) {
      history.events.push({
        timestamp: now,
        date: after.last_reset,
        type: 'status_changed',
        task_id: next.id,
        title: next.title,
        from_status: previous.status,
        to_status: next.status,
        duration: next.duration,
        deadline: next.deadline,
      });
    }

    if (
      previous.title !== next.title ||
      previous.duration !== next.duration ||
      previous.deadline !== next.deadline
    ) {
      history.events.push({
        timestamp: now,
        date: after.last_reset,
        type: 'task_updated',
        task_id: next.id,
        title: next.title,
        to_status: next.status,
        duration: next.duration,
        deadline: next.deadline,
      });
    }
  }

  for (const [id, task] of afterMap) {
    if (beforeMap.has(id)) {
      continue;
    }

    history.events.push({
      timestamp: now,
      date: after.last_reset,
      type: 'task_added',
      task_id: task.id,
      title: task.title,
      to_status: task.status,
      duration: task.duration,
      deadline: task.deadline,
    });
  }
}
